### PR TITLE
FFmpeg 7 support

### PR DESCRIPTION
**Pull request by crsib**
Resolves: #6225

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by saintmatthieu** [comment]
Here's a list of things I think I've understood, please correct me if I'm wrong:
* We support different versions of FFMPEG
* This PR adds support for version 7
* FFMPEG is made of several libraries, including `avcodec`, `avformat`, `avutil`,
* channel layout in version 7 uses a new API, hence a new wrapper is needed and gets injected also in the older `AVUtilImpl`s,
* one same class, `AVChannelLayoutWrapperImpl`, is defined for all versions, but with some `ifdefs` whose value change depending on what `ffmpeg-x.x.x-single-header.h` was included before, and wrapped in different namespaces
* `FFmpegFunctions::Private` is the guy which iterates the supported versions and selects the latest and greatest, if any

That system is quite a feat ... Who was the initiator of this system? Anyone's hand I should shake other than yours, @crsib ? :D

Trying to think of what needs to be tested. The new FIFO is common to all versions, so I suppose that testing with one FFMPEG version would be sufficient for that part.
For the rest, `HAS_AV_CHANNEL_LAYOUT` seems to have some impact, so testing with version before 59 (which FFMPEG is that?) and from 59+.

**Review by saintmatthieu — APPROVED** [review]
Some minor improvement suggestions to please consider, but I couldn't see anything faulty.

**Comment by crsib** [comment]
> We support different versions of FFMPEG

We have to support multiple versions because ... we cannot ship the FFmpeg with Audacity for legal reasons

> This PR adds support for version 7

Yep

> FFMPEG is made of several libraries, including avcodec, avformat, avutil,

We use these three. `avcodec` is a library with all codecs, `avformat` adds support for multiple containers, `avutil` is just used by every FFmpeg library.

> channel layout in version 7 uses a new API, hence a new wrapper is needed and gets injected also in the older AVUtilImpls,

It was introduced in FFmpeg 6, but the old API was killed in 7

> one same class, AVChannelLayoutWrapperImpl, is defined for all versions, but with some ifdefs whose value change depending on what ffmpeg-x.x.x-single-header.h was included before, and wrapped in different namespaces

Yep

> FFmpegFunctions::Private is the guy which iterates the supported versions and selects the latest and greatest, if any

> which FFMPEG is that?

The correspondence between library versions and FFmpeg versions can be found here https://ffmpeg.org/download.html

File: modules/mod-ffmpeg/lib-ffmpeg-support/FifoBuffer.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      void Reset();
+
+      std::vector<char> Data;
+      int WritePosition {};
+      int ReadPosition {};
+   };
+
+   std::deque<Page> mAllocatedPages;
+
+   std::deque<Page*> mActivePages;
```
[the remark this answers is not in the document]
Comment (crsib) [reply modules/mod-ffmpeg/lib-ffmpeg-support/FifoBuffer.h:41]: `queue` is a `dequeue` by default, but I agree that the semantics would be clearer

File: modules/mod-ffmpeg/lib-ffmpeg-support/FifoBuffer.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+   std::vector<Page*> mFreePages;
```
[the remark this answers is not in the document]
Comment (crsib) [reply modules/mod-ffmpeg/lib-ffmpeg-support/FifoBuffer.h:42]: For free pages FILO is as good as FIFO, I don't have any opinion about the container here.

**Pull request merged**